cluster-backup-tool.md

```diff
@@ -0,0 +1,164 @@
+directory structure that contains all backed up cache-containers. Each cache-container then contains it's configured
+cache templates, cache instances and counters. This directory structure is then distributed
+as a archive format, which can be passed to the server in order to initiate a import.
+
+```
+containers/
+containers/<name>
+containers/<name>/cache-container.xml
+containers/<name>/container.properties
+containers/<name>/cache-configs/
+containers/<name>/cache-configs/some-template.xml
+containers/<name>/cache-configs/another-template.xml
+containers/<name>/caches/
+containers/<name>/caches/example-user-cache.dat
+containers/<name>/caches/example-user-cache.xml
+containers/<name>/counters/
+containers/<name>/counters/counters.dat
+
+manifest.properties
+```
+
+The above files will be packaged as a single `.zip` distribution to aid backup/import; this provides compression as well
+as being compatible with both Linux and Windows.
+
+> Utilising the above directory structure simplifies future features such as per container or per cache imports.
+
+## Manifest Properties
+Basic java properties file with key/values which contains cluster wide information.
+
+```java
+version=11.0.0.Final
+cache-containers="default"
+```
+
+The Infinispan version can be used by the Server to quickly determine if a migration to the new server version is possible
+based upon the number of major versions we support migrations across.
+
+## Container Files
+Each container is identified in the directory structure by it's name attribute and contains several sub directories.
+
+### Container Properties
+`container.properties` file that lists all of the configured templates, caches and counters.
+
+```java
+cache-configs="some-template, another-template"
+caches="org.infinispan.CONFIG", "example-user-cache"
+counters="example-counter"
+```
+
+### Container XML
+The container XML can be used to determine if the backup depends on any additional user classes, e.g. serialization marshallers,
+and can check that the server contains these resources on it's classpath, failing fast if they are not present.
+
+### Template files
+Each defined template is represented by a `<template-name>.xml` file.
+
+### Cache Files
+A cache backup consists of a `<cache-name>.xml` and a `<cache-name>.dat` file. The `.xml` file contains
+the cache configuration and is used to create the initial empty cache. The `.dat` file is a batch file containing the cache
+content and is read by the server in order to restore entries to the cache. Each line in the file corresponds to a cache
+entry, with keys and values separated by a single space.
+
+> If the cache is empty during the backup, then no `.dat` file is created.
+
+Example  `.dat` content:
+```java
+hello world
+hola mundo
+```
+
+> How to represent metadata and handle expiration?
+
+#### Storage MediaTypes
+If a cache has a configured key and/or value MediaType, that can be readily represented as a string, e.g. text/plain or json,
```

Comment:
Should keys with spaces and newlines be base64-encoded, too?
Are metadata dropped completely?

Reply:
I just wonder what is the advantage of storing in text format rather than binary. While zipping mostly helps with base64 introduced inflation, I don't see much 'pros'.
```
$ cat /dev/urandom | head -c 65536 > /tmp/r
$ base64 /tmp/r > /tmp/b
$ zip /tmp/b.zip /tmp/b
$ zip /tmp/r.zip /tmp/r
$ ls -l /tmp/b* /tmp/r*
-rw-rw-r--. 1 rvansa rvansa 88534 Jun 19 10:42 /tmp/b
-rw-rw-r--. 1 rvansa rvansa 67444 Jun 19 10:43 /tmp/b.zip
-rw-rw-r--. 1 rvansa rvansa 65536 Jun 19 10:42 /tmp/r
-rw-rw-r--. 1 rvansa rvansa 65706 Jun 19 10:43 /tmp/r.zip
```

Reply:
> You could not tell when it's plaintext and when it's base64...

The plan originally was to rely on the cache configuration to determine this.


> Are metadata dropped completely?

No, I mentioned it above but forgot to detail it's representation. The metadata will have to be serialized and represented accordingly. The open question I have myself is how to handle expiration and max-idle? On backup restore, do we just set the entry creation and last used times to the time of the restore.

> I just wonder what is the advantage of storing in text format rather than binary.

Thanks for mentioning this ... there is none AFAICT. This is a hangover from my initial design where I was going to piggyback on the CLI syntax. Storing in binary format is the way to go :+1:

Reply:
> The open question I have myself is how to handle expiration and max-idle? On backup restore, do we just set the entry creation and last used times to the time of the restore.

There's already a precedent in stores, isn't it? (I just don't remember what that is) What you suggest is simpler but obviously wrong - you should store these timestamps in metadata as is and not change then during restore.

Reply:
> There's already a precedent in stores, isn't it?

Yes, the original values are respected.

> you should store these timestamps in metadata as is and not change then during restore

I'm imagining a scenario where a user has created a backup of a cache, totalling ~ 1Gb. They then restore the backup and have 0 entries in the cache because they have all expired. We could make this behaviour configurable during the import by adding a param, e.g. `/v2/sever?action=import&resetTimestamps=true`.

Reply:
A wise man (more than) once said "less knobs".

If the data is set with expiration, it was set so for a reason - it's not relevant anymore or it even should not be accessible (e.g. here as part as Covid tracking you can give the health officer access to your cell phone location history; all data should be wiped out shortly after the session, though). Cluster failure & backup restore is a technical detail that does nothing to do with the data relevancy.

Reply:
I agree, if it has expiration metadata and is now expired - just don't reinstate that data.